### PR TITLE
Add time and size based message ID deduplication feature

### LINKED ISSUE #207: Message deduplication should have limited time window
Keeping all the message IDs in memory to check for the possible duplicates (given that the deduplication feature is enabled), might result in way too much memory being used for the large number of messages. Instead,  the number of cached IDs should be limited either by the message's amount or the time range, e.g. 1 hour or 1 day.

### PR DESCRIPTION
**Pull request by spetz**
Close #207

### CODE REVIEW

**Review by BartoszCiesla — APPROVED** [review]

File: server/src/streaming/deduplication/message_deduplicator.rs [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    async fn message_deduplicator_should_evict_identifiers_after_given_time_to_live() {
+        let max_entries = 1000;
+        let ttl = Duration::from_millis(10);
+        let deduplicator = MessageDeduplicator::new(Some(max_entries), Some(ttl));
+        for i in 0..max_entries {
+            assert!(deduplicator.try_insert(i as u128).await);
+            assert!(deduplicator.exists(i as u128));
+        }
+
+        sleep(ttl).await;
```
Comment (hubcio) [inline server/src/streaming/deduplication/message_deduplicator.rs]: sleep a tiny bit longer than ttl

File: server/src/streaming/partitions/partition.rs [hunk 2]
```diff
@@ -96,8 +98,22 @@ impl Partition {
...
+                true => Some(MessageDeduplicator::new(
+                    if config.message_deduplication.max_entries > 0 {
+                        Some(config.message_deduplication.max_entries)
+                    } else {
+                        None
+                    },
+                    {
+                        let expiry_secs = config.message_deduplication.expiry.as_secs();
+                        if expiry_secs > 0 {
+                            Some(Duration::from_secs(expiry_secs as u64))
```
Comment (hubcio) [inline server/src/streaming/partitions/partition.rs]: IggyDuration has method as_secs()

File: server/src/streaming/deduplication/message_deduplicator.rs [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        self.cache.contains_key(&id)
+    }
+
+    /// Inserts the given ID.
+    pub async fn insert(&self, id: u128) {
+        self.cache.insert(id, true).await;
+    }
+
+    /// Tries to insert the given ID, returns false if it already exists.
+    pub async fn try_insert(&self, id: u128) -> bool {
```
Comment (hubcio) [inline server/src/streaming/deduplication/message_deduplicator.rs]: perhaps you could try to implement this with`try_get_with_by_ref`?
https://docs.rs/moka/latest/moka/future/struct.Cache.html#method.try_get_with_by_ref

File: server/src/streaming/deduplication/message_deduplicator.rs [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+use std::time::Duration;
+
+#[derive(Debug)]
+pub struct MessageDeduplicator {
+    cache: Cache<u128, bool>,
+}
+
+impl MessageDeduplicator {
+    /// Creates a new message deduplicator with the given max entries and time to live for each ID.
+    pub fn new(max_entries: Option<u64>, ttl: Option<Duration>) -> Self {
```
Comment (hubcio) [inline server/src/streaming/deduplication/message_deduplicator.rs]: `Option<IggyDuration>`

File: server/src/streaming/deduplication/message_deduplicator.rs (hunk 1 above)
Comment (spetz) [reply server/src/streaming/deduplication/message_deduplicator.rs]: Fixed

File: server/src/streaming/partitions/partition.rs (hunk 2 above)
Comment (spetz) [reply server/src/streaming/partitions/partition.rs]: Passing the `IggyDuration` instead

File: server/src/streaming/deduplication/message_deduplicator.rs (hunk 3 above)
Comment (spetz) [reply server/src/streaming/deduplication/message_deduplicator.rs]: Done

**Review by numinnex — APPROVED** [review]

**Pull request merged**